### DISCUSSION TITLE
Show docs for item under cursor - nothing occurs or pops up after SPC + k

### DISCUSSION THREAD

**Discussion by rpherman** [helix-editor/helix#5230] [General]
I was getting a "Language server not active for current buffer". This was resolved on my Windows box of the build-from-source helix with the symlink and putting it manually in my system path.

Rust LSP seems to be configured correctly, but when I hit SPC then 'k' for 'show docs for item under cursor', nothing happens.
Am I missing docs or some other link? Thanks! 

```
Configured language server: rust-analyzer
Binary for language server: E:\helix\rust-analyzer.exe
Configured debug adapter: lldb-vscode
Binary for debug adapter: Not found in $PATH
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```
EDIT: I don't know how to add labels to discussions in Github discussions.

**Comment by kirawi** [comment]
What does it say when you run `hx -v`, open the Rust file, and `:log-open`?

**Reply by rpherman** [reply]
So I ran "hx -v rph.rs" and the log contained 786 lines worth of pretty much these first dozen or so lines.. The timestamp delineates each message:

`2022-12-20T11:30:00.764 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"textDocument":{"codeAction":{"codeActionL    2  2022-12-20T11:30:00.773 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":{"openClose":true,"change":2,"save":{}    3  2022-12-20T11:30:00.773 helix_lsp::transport [INFO] <- {"capabilities":{"callHierarchyProvider":true,"codeActionProvider":{"codeActionKinds":["","quickfix","refac    4  2022-12-20T11:30:00.773 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialized","params":{}}                                                           5  2022-12-20T11:30:00.775 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"method":"window/workDoneProgress/create","params":{"token":"rustAnalyzer/Fetching"    6  2022-12-20T11:30:00.775 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Fetching","value":{"kind":"begin","t    7  2022-12-20T11:30:00.805 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"rust","text":"// A    8  2022-12-20T11:30:00.806 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":null,"id":0}                                                                         9  2022-12-20T11:30:01.001 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Fetching","value":{"kind":"report","   10  2022-12-20T11:30:01.483 helix_lsp::transport [ERROR] err <- "[ERROR project_model::workspace] cyclic deps: helix_view(CrateId(120)) -> helix_tui(CrateId(113)), al   11  2022-12-20T11:30:01.488 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Fetching","value":{"kind":"end"}}}     12  2022-12-20T11:30:01.488 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":1,"method":"window/workDoneProgress/create","params":{"token":"rustAnalyzer/Roots Sca   13  2022-12-20T11:30:01.489 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"begi   14  2022-12-20T11:30:01.489 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"repo   15  2022-12-20T11:30:01.489 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"repo   16  2022-12-20T11:30:01.489 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"repo   17  2022-12-20T11:30:01.489 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"repo   18  2022-12-20T11:30:01.489 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Roots Scanned","value":{"kind":"repo`

**Reply by kirawi** [reply]
You have to wait a bit for the LSP to finish startup. You'll see the spinner at the bottom left disappear completely when it's done.

**Reply by rpherman** [reply]
Wow, thanks! I didn't even see that spinner. I'll wait 5 min. and get back to you.

**Reply by rpherman** [reply]
I think it's working now, but I guess there's nothing to autocomplete? Here are the last few lines of the log with 100 percent complete and two attempts to 'SPC k' over a highlighted word with a "result":null and correctly showing their position.

`2022-12-20T11:41:52.206 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Indexing","value":{"kind":"report","percentage":100}}}
2022-12-20T11:41:52.206 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Indexing","value":{"kind":"end"}}}
2022-12-20T11:42:04.121 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/hover","params":{"position":{"character":3,"line":1},"textDocument":{"uri":"file:///E:/helix/rph.rs"}},"id":1}
2022-12-20T11:42:04.122 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":1,"result":null}
2022-12-20T11:42:04.122 helix_lsp::transport [INFO] <- null
`

Do I need to have a full Rust project, or a file is OK?

I used an example Rust snippet from Rust By Example as follows as the file 'rph.rs'. I hovered and marked "struct" character 3, line 1:

```
// A concrete type `A`.
struct A;

// In defining the type `Single`, the first use of `A` is not preceded by `<A>`.
// Therefore, `Single` is a concrete type, and `A` is defined as above.
struct Single(A);
//            ^ Here is `Single`s first use of the type `A`.

// Here, `<T>` precedes the first use of `T`, so `SingleGen` is a generic type.
// Because the type parameter `T` is generic, it could be anything, including
// the concrete type `A` defined at the top.
struct SingleGen<T>(T);

fn main() {
    // `Single` is concrete and explicitly takes `A`.
    let _s = Single(A);
    
    // Create a variable `_char` of type `SingleGen<char>`
    // and give it the value `SingleGen('a')`.
    // Here, `SingleGen` has a type parameter explicitly specified.
    let _char: SingleGen<char> = SingleGen('a');

    // `SingleGen` can also have a type parameter implicitly specified:
    let _t    = SingleGen(A); // Uses `A` defined at the top.
    let _i32  = SingleGen(6); // Uses `i32`.
    let _char = SingleGen('a'); // Uses `char`.
}

```

**Reply by the-mikedavis** [reply]
rust-analyzer doesn't work out-of-the-box for non-workspace projects (projects not built by cargo). It needs to understand how to build the file(s): https://rust-analyzer.github.io/manual.html#non-cargo-based-projects

**Reply by rpherman** [reply]
Oh, ok. I'll try building a project with cargo and try from there. Thanks @the-mikedavis !

**Reply by rpherman** [reply]
@the-mikedavis That was the answer! I can't see how to mark it on this reply as the answer.
It works as long as I create a rust project via cargo. I tried it with Zig too, and it works as well as long as the project is created in Zig's manner. Thanks so much!

**Comment by rpherman** [comment]
I did run "hx --health rust" to check:

```
E:\helix>hx --health rust
Configured language server: rust-analyzer
Binary for language server: E:\helix\rust-analyzer.exe
Configured debug adapter: lldb-vscode
Binary for debug adapter: Not found in $PATH
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```
